File: c/day6.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "aoc.h"

#define DAY 6

#define MAX_STAGES 9
#define MAX_INITIAL_NUMBERS 1000
#define MAX_LINE_LENGTH 1000
#define MAX_LINES 2

void part_a();
void part_b();

char** lines;
int line_count = 0;
int numbers[MAX_INITIAL_NUMBERS];
long long stages[MAX_STAGES];

void main()
{
	start_day(DAY);

	lines = read_input(DAY, "../", MAX_LINES, MAX_LINE_LENGTH, &line_count);
	if (lines == NULL)
	{
		perror("read_input()");
		return;
	}

    rtrim(lines[0]);

    for (int ix = 0; ix < MAX_INITIAL_NUMBERS; ix++)
    {
        numbers[ix] = -1;
    }

    char delim[] = ",";
    char *ptr = strtok(lines[0], delim);
    int ix = 0;
    while (ptr != NULL)
    {
        numbers[ix++] = atoi(ptr);
        ptr = strtok(NULL, delim);
    }

    part_a();
    part_b();

	free_input(lines, line_count);
}

long long breed(int days)
{
    for (int s = 0; s < MAX_STAGES; s++)
    {
        stages[s] = 0;
    }

    for (int ix = 0; ix < MAX_INITIAL_NUMBERS; ix++)
    {
        if (numbers[ix] < 0)
        {
            continue;
        }

        stages[numbers[ix]]++;
    }

    for (int i = 0; i < days; i++)
    {
        long long pregnant = stages[0];
        for (int s = 1; s < MAX_STAGES; s++)
        {
            stages[s - 1] = stages[s];
            stages[s] = 0;
        }

        stages[6] += pregnant;
        stages[8] += pregnant;
    }

    long long sum = 0;
    for (int s = 0; s < MAX_STAGES; s++)
    {
        sum += stages[s];
    }

    return sum;
}

void part_a()
{
	start_part_a();

    long long result = breed(80);

	show_answer(result);
}

void part_b()
{
	start_part_b();

    long long result = breed(256);

	show_answer(result);
}

File: day6.py
from aoc import Aoc
import itertools
import math
import re
import sys

# Day 6
# https://adventofcode.com/2021

class Day6Solution(Aoc):

    def Run(self):
        self.StartDay(6)
        self.ReadInput()
        self.PartA()
        self.PartB()

    def Test(self):
        self.StartDay(6)

        goal = self.TestDataA()
        self.PartA()
        self.Assert(self.GetAnswerA(), goal)

        goal = self.TestDataB()
        self.PartB()
        self.Assert(self.GetAnswerB(), goal)

    def TestDataA(self):
        self.inputdata.clear()
        testdata = \
        """
		3,4,3,1,2
        """
        self.inputdata = [line.strip() for line in testdata.strip().split("\n")]
        return 5934

    def TestDataB(self):
        self.inputdata.clear()
        self.TestDataA()
        return 26984457539

    def Breed(self, days:int) -> int:
        numbers = [int(x) for x in self.inputdata[0].split(",")]
        stages = [len([x for x in numbers if x == d]) for d in range(8 + 1)]
        for day in range(days):
        	pregnant = stages[0]
        	for x in range(1, len(stages)):
        		stages[x - 1] = stages[x]
        		stages[x] = 0
        	stages[6] += pregnant
        	stages[8] += pregnant

        return sum(stages)

    def PartA(self):
        self.StartPartA()

        answer = self.Breed(80)

        self.ShowAnswer(answer)

    def PartB(self):
        self.StartPartB()

        answer = self.Breed(256)

        self.ShowAnswer(answer)


if __name__ == "__main__":
    solution = Day6Solution()
    if len(sys.argv) >= 2 and sys.argv[1] == "test":
        solution.Test()
    else:
        solution.Run()

# Template Version 1.3

